Added blog post on nesting typegen files (#87)
statelyai/eng-blog: content/posts/2022-06-13-nesting-typegen-files.mdx

---
title: "Nesting XState typegen files"
description: What is file nesting in VS Code and how to enable it for XState's generated type files.
tags:
  - stately
  - extension
  - open source
  - xstate
  - vscode
author:
  - Anders Bech Mellson
publishedAt: "2022-06-13"
ogImage: "https://stately.ai/blog/2022-06-13-nesting-typegen-files.png"
---

Our latest [XState VS Code extension](https://marketplace.visualstudio.com/items?itemName=statelyai.stately-vscode) update has made it easy to enable file nesting for typegen files. But what is it?

## Explorer File Nesting

VS Code [introduced file nesting](https://code.visualstudio.com/updates/v1_67#_explorer-file-nesting) in version 1.67 back in April of 2022. This feature is elegant as it allows you to group related files in the explorer view of VS Code.

Our friend; Erik Rasmussen quickly found that this feature works well with the generated type files for XState. Thanks to Erik for his great idea!

<Tweet id="1522611081528573956" />

## Show me

> This is how the explorer in VS Code looks without file nesting enabled when you have a machine source file alongside a generated type file for that machine.

Reviewer: ```suggestion
The image below shows the explorer in VS Code without file nesting enabled; you have a machine source file alongside a generated type file for that machine.
```

> ![VS Code explorer view with an XState machine alongside the types generated for it.](./images/without-typegen.png)

Reviewer: ```suggestion
![The source file in the VS Code explorer called coolMachine.ts is listed above the typegen file called coolMachine.typegen.ts.](./images/without-typegen.png)
```

> And this is how it looks after enabling file nesting for typegen files. Now it allows you to hide the generated file behind the source if you want to.

Reviewer: ```suggestion
After enabling file nesting for typegen files, VS Code explorer allows you to nest and collapse the generated file under the source file.
```

> ![Animated GIF showing how it looks when you toggle nesting of a generated type file.](./images/with-typegen.gif)

Reviewer: ```suggestion
![Pressing the arrow icon toggle alongside the source file toggles the nesting of the generated type file.](./images/with-typegen.gif)
```

> If you’re using our VS Code extension and haven’t manually enabled file nesting for typegen files, you will see a helper prompt allowing you to enable file nesting with 1 click.

Reviewer: Does this helper prompt appear when you open VS Code for the first time after update? Or when you perform a particular action with the extension?

<Tweet id="1522611081528573956" />

## Show me

The image below shows the explorer in VS Code without file nesting enabled; you have a machine source file alongside a generated type file for that machine.

![The source file in the VS Code explorer called coolMachine.ts is listed above the typegen file called coolMachine.typegen.ts.](./images/without-typegen.png)

After enabling file nesting for typegen files, VS Code explorer allows you to nest and collapse the generated file under the source file.

![Pressing the arrow icon toggle alongside the source file toggles the nesting of the generated type file.](./images/with-typegen.gif)

### Enable nesting

If you’re using our VS Code extension and haven’t manually enabled file nesting for typegen files, you will see a helper prompt allowing you to enable file nesting with a click.
The prompt will show when you open a file with a machine.
If you miss the prompt, you can enable or disable file nesting as described in the next section.

### Disable nesting

You can disable, or enable, the file nesting by using the `Xstate: Nest Typegen Files` toggle in your VS Code settings.

> ![The settings view in VS Code showing the settings specific for XState.](./images/xstate-vscode-settings.png)

Reviewer: ```suggestion
![The Extensions settings view in VS Code has XState settings including the option to enable nesting the generated typegen files.](./images/xstate-vscode-settings.png)
```
